### ISSUE TITLE
RFC: New top-level API for Swift 3.1

### ISSUE THREAD

**Issue by JohnSundell** [JohnSundell/Unbox#160]
Hi Unbox users!

Swift 3.1 is coming with Xcode 8.3, and includes the exciting new capability of defining same-type constraints for extensions on generics. With this, I'm planning to change the top-level unboxing API to become (IMO) a lot simpler and nicer.

Right now, unboxing a model is done by calling the global `unbox()` function that comes in a few overloads, that each have different input and output. Unboxing a `Dictionary`  into a `Model` looks like this:

```swift
let model: Model = try unbox(dictionary: dictionary)
```

Now, with Swift 3.1, an extension can be made on `Dictionary` with constraints that matches JSON, like this:

```swift
extension Dictionary where Key == String, Value == Any {
    func unbox<T: Unboxable>() throws -> T
}
```

This would change the way the API is used to this:

```swift
let model: Model = try dictionary.unbox()
```

Similarily, we can create extensions on `Array` and `Data`, giving us a very nice, unified API:

```swift
let model: Model = try dictionary.unbox()
let model: Model = try data.unbox()
let model: [Model] = try array.unbox()
let model: [Model] = try data.unbox()
```

What do you think? Do you think the suggested change would make Unbox nicer and easier to use? Let me know by posting a reaction or in the comments. Thanks! 🚀

**Comment by harlanhaskins** [comment]
I think it'd make it much easier given how many people expect autocompletion after pressing `.`.

**Comment by robertherdzik** [comment]
Cool! this face of Unbox looks way cleaner 😊 💥

**Comment by icanzilb** [comment]
I can see the benefits of the current syntax though - you always use the same function + XCode shows you an autocomplete with all possible types you can unbox 👍 

The new syntax feels cleaner

**Comment by thomaspaulmann** [comment]
Looks promising to me, it feels super clean and really natural to use, especially with the type inference of Swift 👍 
If you need any help to implement the new API, pleas lmk 💪

**Comment by eliburke** [comment]
👍🏻here.

**Comment by azimin** [comment]
Yes, it looks cleaner and easier to call 👍💶

**Comment by BalestraPatrick** [comment]
Big fan of the new syntax. It's shorter and easier to remember.

**Comment by mislavjavor** [comment]
I *do* think the new syntax is cool, but not sure it's worth breaking the API. Maybe keep both versions without deprecating the old syntax?

**Comment by christopherpow** [comment]
Love it!

**Comment by gperdomor** [comment]
@johannth also update travis to include linux tests please

**Comment by JohnSundell** [comment]
Thanks everybody for your input on this! 🎉

Since the new API has gotten such a positive reaction, and since Swift 3.1 is now out - I'm going to go ahead and implement this 🚀 I'm going to deprecate the old API, so that it still works, but a warning with instructions on how to upgrade to the new API will be emitted. The old API will be removed in Unbox 4.0.

**Comment by aventurella** [comment]
I'm sure this is too late, but I figured I'd add my 2c for what it's worth.

While I agree that it's a nice take on taking advantage of a new feature, I have to say I prefer the explicit nature of the original style better:

```
import Unbox
let model: Model = try unbox(dictionary: dictionary)
```

`T` is not magically imbued with anything and the call site makes it pretty clear what exactly is happening:

```
1) Someone imported this lib called Unbox
2) that act of that import provides this function by the same name it's clearly related.
```

vs the new style:

```
import Unbox
let model: Model = try dictionary.unbox()
```

Because  ANY other code in the application could have this extension rule:

```
extension Dictionary where Key == String, Value == Any {
    func unbox<T: Unboxable>() throws -> T
}
```

you run the risk of name conflicts and then who wins?  I tend to think of extensions that add methods like that to be reserved for the application author: AKA it's their maintenance problem to understand what methods they have added towards their ends.  Libraries on the other hand should be good citizens. Since they have 0 knowledge of what the application author has done they shouldn't go around adding extensions to things lest they get in the way.

Bottom line is I tend to err on staying out of the application authors way at all times. They need to be able to make their own decisions about how they add extensions and what they call them. The second a library collides with what the author needs to do, that library will be tainted.

Keeping the top level `unbox` function explicit pretty much ensures you **NEVER** get in the way. If a name **DOES** collied, the application author simply prefixes it with a namespace: `Unbox.unbox` to clarify the intention.

**Comment by JohnSundell** [comment]
@aventurella I've only done some preparations for this new API so far, so no worries, there's still time to discuss it 👍 

While I see your point and I totally agree libraries should always be good citizens - in practice I don't think this particular case will be much of an issue. Since Swift supports method overloading, the **only** way that this new API could conflict with application code or another library is if a method is added with the exact same signature on `Dictionary` or `Data`.

I wouldn't add an API like this for example:

```swift
extension Dictionary {
    func asString() -> String
}
```

since that has a big potential for conflict, giving the naming and return type of that method. But in the case of this new API, it both relies on the `Unboxable` type as a generic constraint, and is called `unbox()`, both of which I think is specific enough not to cause conflicts.

So even if the application code adds something like this:

```swift
extension Dictionary where Key == String, Value == Any {
    func unbox() -> String
}

we won't have any conflicts.

**Comment by aventurella** [comment]
Ya, you are right. It won't conflict. So that just leaves the: `the act of importing changes things implicitly` AKA it introduces the type extensions.  This is simply solved with strong documentation, and the README for the projects I think covers that, so I'm not worried there.

My personal taste is that I like to be explicit about things. So I definitely favor `unbox(...)` vs `someVar.unbox()`. The latter implies I actually read the docs and understand this method is an extension method =)

Again that's just my personal taste. I use unbox in everything, and will continue to use it in everything either way you decide =)

**Comment by JohnSundell** [comment]
It's good input 👍 both ways have pros & cons, for sure.

The reason I'm still in favor of implementing the new API is that `unbox()` is an action that you take **on the** dictionary/data you are unboxing. The dictionary/data is not merely input into some other kind of operation, it's the actual target of your operation. That's why I think it makes sense to have it as an extension medhod on `Dictionary` and `Data`. Just like you wouldn't do `uppercase(string)`, you do `string.uppercased()`.

I've actually been wanting to implement this API for quite some time, but it's only now with Swift 3.1 that it's possible 😄

**Comment by clayellis** [comment]
The `uppercase(string)` vs `string.uppercased()` is a great example of why this is useful. I remember when the change was made from `for (index, item) in enumerate(items) ...` to `for (index, item) in items.enumerated() ...` — it makes more sense to not use a global function and actually, in my opinion, increases discoverability of the `enumerated()` method. Granted, those using `Unbox` won't just be "discovering" the functionality on the fly. I think the point still stands, though.

This raises a question as to how the extension should be named. From what I understand, the convention is to add an "`ed`" to methods which do not actually mutate the receiver but instead return a transformed (loosely using the word here) version. 

So should it be `dictionary.unboxed()` rather than `dictionary.unbox()`? The [Swift API Design Guidelines](https://swift.org/documentation/api-design-guidelines/) sure seem to favor the `unboxed()` variant:

> Name Mutating/nonmutating method pairs consistently. A mutating method will often have a nonmutating variant with similar semantics, but that returns a new value rather than updating an instance in-place.

> When the operation is naturally described by a verb, use the verb’s imperative for the mutating method and apply the “ed” or “ing” suffix to name its nonmutating counterpart.

I feel like it should be `unboxed()`. What do you think, @JohnSundell?

**Comment by JohnSundell** [comment]
@clayellis I think you're totally right! Let's go with `unboxed()` 👍 Thanks for bringing this up! 🚀

**Comment by gperdomor** [comment]
The wrap project will follow the same pattern? something.wraped() ?

**Comment by JohnSundell** [comment]
@gperdomor No, since there's no way of adding a `wrapped()` method to all types currently (and I don't want to limit Wrap to only types conforming to a protocol such as `Wrappable`, I want to be usable with any instance of any type) 🙂

**Comment by JohnSundell** [comment]
I've now started the implementation of this (along with the rest of Unbox 3.0 changes) here: https://github.com/JohnSundell/Unbox/pull/175 🚀

**Comment by eonist** [comment]
@JohnSundell Swift 4 brings Decodable and Encodable. Which means no more custom types for every class you make. Whats your thoughts on this?
